constraints/memory.air

```diff
@@ -2,13 +2,253 @@
+ev memory_chiplet_constraints_all_rows([memory[15]]) {
+    let is_read = memory[0];        # Read(1)/Write(0) selector
+    let is_word_access = memory[1]; # Element(0)/Word(1) access selector
+    let idx0 = memory[4];           # Element index bit 0
+    let idx1 = memory[5];           # Element index bit 1
+
+    # Binary column constraints
+    enf enforce_binary_columns([is_read, is_word_access, idx0, idx1]);
+}
+
+# Enforces memory state transition constraints for all rows except the final row
+# Includes monotonicity, value consistency, and proper read/write semantics
+ev memory_chiplet_constraints_all_rows_except_last([memory[15]]) {
+    let is_read = memory[0];        # Read(1)/Write(0) selector
+    let is_word_access = memory[1]; # Element(0)/Word(1) access selector
+    let ctx = memory[2];            # Context ID
+    let addr = memory[3];           # Memory address
+    let idx0 = memory[4];           # Element index bit 0
+    let idx1 = memory[5];           # Element index bit 1
+    let clk = memory[6];            # Clock cycle
+    let v0 = memory[7];             # Memory value 0
+    let v1 = memory[8];             # Memory value 1
+    let v2 = memory[9];             # Memory value 2
+    let v3 = memory[10];            # Memory value 3
+    let d0 = memory[11];            # Context delta
+    let d1 = memory[12];            # Address delta
+    let d_inv = memory[13];         # Delta inverse
+    let f_scw = memory[14];         # Same context/word flag
+
+    # Delta inverse constraints
+    enf enforce_d_inv([ctx, addr, clk, d0, d1, d_inv]);
+
+    # Context/address/clock delta constraints
+    enf enforce_delta([ctx, addr, clk, d0, d1, d_inv]);
+
+    # Same context/word flag constraints
+    enf enforce_flag_same_context_and_word([ctx, addr, d_inv, f_scw]);
+
+    # Same context/word/addr/clock access constraints
+    enf enforce_same_context_word_addr_and_clock([is_read, ctx, clk, d_inv]);
+
+    # Memory value constraints
+    enf enforce_values_consistency([is_read, is_word_access, idx0, idx1, v0, v1, v2, v3, f_scw]);
+}
+
+##########################################################################################
+# HELPER EVALUATORS
+##########################################################################################
+
+# Constrains binary columns for selectors and indices
+ev enforce_binary_columns([is_read, is_word_access, idx0, idx1]) {
+    # Read/write selector must be binary
+    enf is_binary([is_read]);
+
+    # Element/word access selector must be binary
+    enf is_binary([is_word_access]);
+
+    # Index bit 0 must be binary
+    enf is_binary([idx0]);
+
+    # Index bit 1 must be binary
+    enf is_binary([idx1]);
+}
+
+# Constrains the delta inverse column
+ev enforce_d_inv([ctx, addr, clk, d0, d1, d_inv]) {
+    let ctx_delta = ctx' - ctx;
+    let addr_delta = addr' - addr;
+    let n0 = ctx_delta * d_inv';
+    let n1 = addr_delta * d_inv';
```

Comment:
In the docs, the constraint for n0 binary is: fmem_nl*(n0^2 - n0) = 0. Here we use binary_constraint(n0) = 0 which is equivalent, but the degree analysis in docs shows this as degree 7. The constraint seems to be missing the fmem_nl flag for when the constraint should apply.